### PR TITLE
Add gRPC support to RealtimeAPI kind

### LINKED ISSUE #1056: Add gRPC support to RealtimeAPI kind
#### Description

https://grpc.io/docs/guides/concepts/
https://blog.feathersjs.com/http-vs-websockets-a-performance-comparison-da2533f13a77

1. For best performance, the protocol would be customizable in the API spec (REST vs gRPC). A field called `protocol` would be added to the `predictor` section.
1. User is responsible for providing a protobuf file in the API spec as well. Cortex would automatically generate the server files from the protobuf.
1. Both REST and gRPC cannot be served for the same API.
1. Istio doesn't have to be changed - it's got native support for HTTP 2.0. We probably need to change the service mode of APIs to headless mode.
1. The protobuf would only support a single method/service - `predict`. Its input/output is defined by the user. We don't provide a default protobuf.

Example of python predictor for gRPC:

```python
# UserProvidedServicer provides an implementation of the methods of the RouteGuide service.
class PythonPredictor(user_pb2_grpc.UserServicer):
  def start(self, config):
    self.config = config
  def predict(self, request_iterator, context):
    prev_notes = []
    for new_note in request_iterator:
      for prev_note in prev_notes:
        if prev_note.location == new_note.location:
          yield prev_note
      prev_notes.append(new_note)
```
We don't necessarily need (or can) to ask the user to subclass the generated server files - we can wrap this class instead. This way, the class could look like

```python
class PythonPredictor:
  def __init__(self, config):
    self.config = config
  def predict(self, request_iterator, context):
    prev_notes = []
    for new_note in request_iterator:
      for prev_note in prev_notes:
        if prev_note.location == new_note.location:
          yield prev_note
      prev_notes.append(new_note)
```

Creating a gRPC server is straightforward. We'll need multiple servers as dictated by `processes_per_replica` field.

```python
def serve():
  server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
  user_pb2_grpc.add_UserServices_to_server(
      PythonPredictor(), server)
  server.add_insecure_port('[::]:50051')
  server.start()
  server.wait_for_termination()
```

Example API spec:

```yaml
- name: iris-classifier
  kind: RealtimeAPI
  predictor:
    type: python
    path: predictor.py
    protocol: grpc/rest
    protobuf: predictor.proto
  compute:
    cpu: 0.2
    mem: 200M
```

To parse proto files in Go: https://github.com/tallstoat/pbparser

#### Motivation

Because it has lower latency and higher throughput than REST. It's already well known within the community:
* https://medium.com/@EmperorRXF/evaluating-performance-of-rest-vs-grpc-1b8bdf0b22da

#### Additional context

Maybe use something like Linkerd *(maybe this isn't needed)* for load-balancing gRPC requests:
* https://kubernetes.io/blog/2018/11/07/grpc-load-balancing-on-kubernetes-without-tears/

### PR DESCRIPTION
**Pull request by RobertLucian**
Closes #1056.

---

checklist:

- [ ] run `make test` and `make lint`
- [ ] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)
- [ ] update examples
- [ ] update docs and add any new files to `summary.md` (view in [gitbook](https://cortex-labs.gitbook.io/staging/-MOmCGMADSRNQahK3Kox/) after merging)

### CODE REVIEW

**Review by vishalbollu — COMMENTED** [review]
Partial review progress

File: pkg/cortex/serve/init/bootloader.sh [hunk 1]
```diff
@@ -142,18 +142,34 @@ create_s6_service_from_file() {
 
 # prepare webserver
 if [ "$CORTEX_KIND" = "RealtimeAPI" ]; then
+    if [ $CORTEX_SERVING_PROTOCOL = "http" ]; then
+        mkdir /run/uvicorn
+    fi
+
+    if [ $CORTEX_SERVING_PROTOCOL = "grpc" ]; then
+        /opt/conda/envs/env/bin/python -m grpc_tools.protoc --proto_path=$CORTEX_PROJECT_DIR --python_out=$CORTEX_PYTHON_PATH --grpc_python_out=$CORTEX_PYTHON_PATH $CORTEX_PROTOBUF_FILE
```
Comment (vishalbollu) [inline pkg/cortex/serve/init/bootloader.sh:150]: should these files be created in a separate directory so they don't potentially override any existing files in /mnt/project?

File: pkg/cortex/serve/init/bootloader.sh [hunk 2]
```diff
@@ -142,18 +142,34 @@ create_s6_service_from_file() {
...
+        fi
     done
 
     # generate nginx conf
     /opt/conda/envs/env/bin/python -c 'from cortex_internal.lib import util; import os; generated = util.render_jinja_template("/src/cortex/serve/nginx.conf.j2", os.environ); print(generated);' > /run/nginx.conf
 
     create_s6_service "py_init" "cd /mnt/project && exec /opt/conda/envs/env/bin/python /src/cortex/serve/init/script.py"
     create_s6_service "nginx" "exec nginx -c /run/nginx.conf"
+
+    touch /mnt/workspace/api_readiness.txt
```
Comment (RobertLucian) [inline pkg/cortex/serve/init/bootloader.sh:172]: Check if we can remove this

File: pkg/cortex/serve/start/server_grpc.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+                if not cron.is_alive():
+                    os.kill(os.getpid(), signal.SIGQUIT)
+            time.sleep(1)
+
+    threading.Thread(target=check_if_crons_have_failed, daemon=True).start()
+
+    ServicerClass = get_servicer_from_module(module_proto_pb2_grpc)
+
+    class PredictorServicer(ServicerClass):
+        def __init__(self, config: Dict[str, Any]):
```
Comment (RobertLucian) [inline pkg/cortex/serve/start/server_grpc.py:161]: Just pass in the predictor fn args, impl and api.

File: pkg/cortex/serve/start/server_grpc.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    SERVICE_NAMES = (
+        module_proto_pb2.DESCRIPTOR.services_by_name[service_name].full_name,
+        reflection.SERVICE_NAME,
+    )
+    reflection.enable_server_reflection(SERVICE_NAMES, server)
+
+    server.add_insecure_port(address)
+    server.start()
+
+    time_to_wait = 5.0
```
Comment (RobertLucian) [inline pkg/cortex/serve/start/server_grpc.py:223]: Just do a damn sleep(5).

File: pkg/lib/configreader/errors.go [hunk 5]
```diff
@@ -35,10 +35,12 @@ const (
 	ErrLeadingWhitespace             = "configreader.leading_whitespace"
 	ErrTrailingWhitespace            = "configreader.trailing_whitespace"
 	ErrAlphaNumericDashUnderscore    = "configreader.alpha_numeric_dash_underscore"
+	ErrAlphaNumericUnderscore        = "configreader.alpha_numeric_underscore"
```
Comment (RobertLucian) [inline pkg/lib/configreader/errors.go:38]: It's dot underscore.

File: pkg/operator/resources/errors.go [hunk 6]
```diff
@@ -27,14 +27,16 @@ import (
...
+	ErrOperationIsOnlySupportedForKind                          = "resources.operation_is_only_supported_for_kind"
+	ErrAPINotDeployed                                           = "resources.api_not_deployed"
+	ErrAPIIDNotFound                                            = "resources.api_id_not_found"
+	ErrCannotChangeTypeOfDeployedAPI                            = "resources.cannot_change_kind_of_deployed_api"
+	ErrCannotChangeAPIServingProtocolWhenTrafficSplitterIsInUse = "resources.cannot_change_api_serving_protocol_when_traffic_splitter_is_in_use"
+	ErrNoAvailableNodeComputeLimit                              = "resources.no_available_node_compute_limit"
+	ErrJobIDRequired                                            = "resources.job_id_required"
+	ErrRealtimeAPIUsedByTrafficSplitter                         = "resources.realtime_api_used_by_traffic_splitter"
+	ErrAPIsNotDeployed                                          = "resources.apis_not_deployed"
+	ErrGrpcAPINotSupportedForTrafficSplitter                    = "resources.grpc_api_not_supported_for_traffic_splitter"
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply pkg/operator/resources/errors.go:39]: `ErrGRPCNotSupportedForTrafficSplitter`

File: pkg/operator/resources/errors.go [hunk 7]
```diff
@@ -72,6 +74,13 @@ func ErrorCannotChangeKindOfDeployedAPI(name string, newKind, prevKind userconfi
 	})
 }
 
+func ErrorCannotChangeAPIServingProtocolWhenTrafficSplitterIsInUse(protocolChangingAPIName string, trafficSplitters []string) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrCannotChangeAPIServingProtocolWhenTrafficSplitterIsInUse,
```
Comment (RobertLucian) [inline pkg/operator/resources/errors.go:79]: ` ErrCannotChangeProtocolWhenUsedByTrafficSplitter`

File: pkg/operator/resources/realtimeapi/k8s_specs.go [hunk 8]
```diff
@@ -122,11 +135,12 @@ func pythonAPISpec(api *spec.API, prevDeployment *kapps.Deployment) *kapps.Deplo
...
-				"apiName":        api.Name,
-				"apiKind":        api.Kind.String(),
-				"deploymentID":   api.DeploymentID,
-				"predictorID":    api.PredictorID,
-				"cortex.dev/api": "true",
+				"apiName":         api.Name,
+				"apiKind":         api.Kind.String(),
+				"deploymentID":    api.DeploymentID,
+				"predictorID":     api.PredictorID,
+				"servingProtocol": servingProtocol,
```
Comment (RobertLucian) [inline pkg/operator/resources/realtimeapi/k8s_specs.go:142]: Remove this label if it's not used anywhere.

File: pkg/operator/resources/resources.go [hunk 9]
```diff
@@ -127,7 +127,15 @@ func Deploy(projectBytes []byte, configFileName string, configBytes []byte, forc
 	results := make([]schema.DeployResult, 0, len(apiConfigs))
 	for i := range apiConfigs {
 		apiConfig := apiConfigs[i]
-		api, msg, err := UpdateAPI(&apiConfig, projectID, force)
+
+		protobufBytes := []byte{}
+		if apiConfig.Predictor != nil && apiConfig.Predictor.ProtobufPath != nil {
+			protobufBytes, err = projectFiles.GetFile(*apiConfig.Predictor.ProtobufPath)
```
Comment (RobertLucian) [inline pkg/operator/resources/resources.go:133]: Remove protobytes.

File: pkg/operator/resources/trafficsplitter/k8s_specs.go [hunk 10]
```diff
@@ -34,7 +34,7 @@ func virtualServiceSpec(trafficSplitter *spec.API) *istioclientnetworking.Virtua
 		Gateways:     []string{"apis-gateway"},
 		Destinations: getTrafficSplitterDestinations(trafficSplitter),
 		ExactPath:    trafficSplitter.Networking.Endpoint,
-		Rewrite:      pointer.String("predict"),
+		Rewrite:      pointer.String("/"),
```
Comment (RobertLucian) [inline pkg/operator/resources/trafficsplitter/k8s_specs.go:37]: This should be reverted.

File: pkg/types/spec/errors.go [hunk 11]
```diff
@@ -43,6 +43,7 @@ const (
 	ErrSpecifyAllOrNone             = "spec.specify_all_or_none"
 	ErrOneOfPrerequisitesNotDefined = "spec.one_of_prerequisites_not_defined"
 	ErrConfigGreaterThanOtherConfig = "spec.config_greater_than_other_config"
+	ErrInvalidFieldValue            = "spec.invalid_field_value"
```
Comment (RobertLucian) [inline pkg/types/spec/errors.go:46]: Remove this

File: test/apis/grpc/iris-classifier-sklearn/trainer.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
+import boto3
+import pickle
+
+from sklearn.datasets import load_iris
+from sklearn.model_selection import train_test_split
+from sklearn.linear_model import LogisticRegression
+
+# Train the model
+
+iris = load_iris()
```
Comment (RobertLucian) [inline test/apis/grpc/iris-classifier-sklearn/trainer.py:10]: Remove this

File: test/apis/grpc/prime-number-generator/README.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
+## Prime number generator
+
+#### Step 1
+
+```bash
+pip install grpc
+python -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. iris_classifier.proto
```
Comment (RobertLucian) [inline test/apis/grpc/prime-number-generator/README.md:7]: Use the correct name for the proto.

**Review by vishalbollu — APPROVED** [review]
